File: 07-browser-automation/golang/providers/registry.go
package providers

import (
	"context"
	"fmt"
	"os"
	"strings"
)

var registry = map[string]Provider{}

func Register(p Provider) {
	registry[p.Name()] = p
}

// ChatWithFallback tries each provider in chain order; moves to the next on any error.
func ChatWithFallback(ctx context.Context, messages []Message, chain []string, assemble func([]Message, Provider) ([]Message, error)) (string, error) {
	var errs []string
	for _, name := range chain {
		p, ok := registry[name]
		if !ok {
			errs = append(errs, name+": not registered")
			continue
		}
		assembled, err := assemble(messages, p)
		if err != nil {
			errs = append(errs, fmt.Sprintf("%s (context assembly): %v", name, err))
			continue
		}
		reply, err := p.Chat(ctx, assembled)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[provider:%s] failed — %v\n", name, err)
			errs = append(errs, fmt.Sprintf("%s: %v", name, err))
			continue
		}
		return reply, nil
	}
	return "", fmt.Errorf("all providers failed:\n%s", strings.Join(errs, "\n"))
}

// StreamWithFallback calls provider.Stream(); on error moves to next provider.
func StreamWithFallback(ctx context.Context, messages []Message, chain []string, assemble func([]Message, Provider) ([]Message, error), onToken func(string)) (string, error) {
	var errs []string
	for _, name := range chain {
		p, ok := registry[name]
		if !ok {
			errs = append(errs, name+": not registered")
			continue
		}
		assembled, err := assemble(messages, p)
		if err != nil {
			errs = append(errs, fmt.Sprintf("%s (context assembly): %v", name, err))
			continue
		}
		reply, err := p.Stream(ctx, assembled, onToken)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[provider:%s] failed — %v\n", name, err)
			errs = append(errs, fmt.Sprintf("%s: %v", name, err))
			continue
		}
		return reply, nil
	}
	return "", fmt.Errorf("all providers failed:\n%s", strings.Join(errs, "\n"))
}

File: 12-observability/nodejs/src/providers/registry.ts
import type { Message, Provider, StreamResult } from './types.ts';
import { assembleContext } from '../context.ts';
import { warn } from '../logger.ts';

// ── Provider registry ────────────────────────────────────────────────────────

const providerRegistry = new Map<string, Provider>();

export function registerProvider(provider: Provider) {
  providerRegistry.set(provider.name, provider);
}

export function getProvider(name: string): Provider | undefined {
  return providerRegistry.get(name);
}

// ── Fallback router ──────────────────────────────────────────────────────────

// Tries each provider in order; moves to the next on any error.
export async function chatWithFallback(
  messages: Message[],
  chain: string[],
): Promise<string> {
  const errors: string[] = [];

  for (const name of chain) {
    const provider = providerRegistry.get(name);
    if (!provider) {
      errors.push(`${name}: not registered`);
      continue;
    }

    const ctx = await assembleContext(messages, provider);
    try {
      return await provider.chat(ctx);
    } catch (err: any) {
      const msg = err?.message ?? String(err);
      warn(`[provider:${name}] failed — ${msg}`);
      errors.push(`${name}: ${msg}`);
    }
  }

  throw new Error(`All providers failed:\n${errors.join('\n')}`);
}

// Streaming variant: calls provider.stream() if available, otherwise falls back
// to provider.chat() with a single synthetic onToken call at the end.
export async function streamWithFallback(
  messages: Message[],
  chain: string[],
  onToken: (token: string) => void,
): Promise<StreamResult> {
  const errors: string[] = [];

  for (const name of chain) {
    const provider = providerRegistry.get(name);
    if (!provider) {
      errors.push(`${name}: not registered`);
      continue;
    }

    const ctx = await assembleContext(messages, provider);
    try {
      if (provider.stream) {
        return await provider.stream(ctx, onToken);
      }
      // Degrade gracefully: no streaming → emit full reply as one token
      const reply = await provider.chat(ctx);
      onToken(reply);
      return { reply };
    } catch (err: any) {
      const msg = err?.message ?? String(err);
      warn(`[provider:${name}] failed — ${msg}`);
      errors.push(`${name}: ${msg}`);
    }
  }

  throw new Error(`All providers failed:\n${errors.join('\n')}`);
}
